src/main/drivers/mco.c

```diff
@@ -27,26 +27,82 @@
 #ifdef USE_MCO
 
 #include "drivers/io.h"
+#include "drivers/mco.h"
 #include "pg/mco.h"
 
-void mcoInit(const mcoConfig_t *mcoConfig)
+#ifdef STM32G4
+
+// Notes
+// - MCO output stability
+// MCO output should not be too high.
+// For example
+// it is required to use DIV4 for SYSCLK = 254MHz (derives 62.5MHz)
+// it is required to use DIV2 for SYSCLK = 170MHz (derives 85MHz)
+//
+// - MCO frequenchy can be more flexible if PLLR is made configurable.
+
+const uint32_t mcoSources[MCO_SOURCE_COUNT] = {
+    RCC_MCO1SOURCE_NOCLOCK,
+    RCC_MCO1SOURCE_SYSCLK,
+    RCC_MCO1SOURCE_HSI,
+    RCC_MCO1SOURCE_HSE,
+    RCC_MCO1SOURCE_PLLCLK, // PLLR on G4
+    RCC_MCO1SOURCE_LSI,
+    RCC_MCO1SOURCE_LSE,
+    RCC_MCO1SOURCE_HSI48,
+};
+
+const uint32_t mcoDividers[MCO_DIVIDER_COUNT] = {
+    RCC_MCO_DIV1,
+    RCC_MCO_DIV2,
+    RCC_MCO_DIV4,
+    RCC_MCO_DIV8,
+    RCC_MCO_DIV16,
+};
+#endif
+
+static void mcoConfigure(MCODevice_e device)
 {
+    const mcoConfig_t *config = mcoConfig(device);
```

Comment:
There should not be any access to the firmware configuration in the driver layer - this should be passed down from the firmware layer, like it originally was. Probably the easiest way to get this done is to integrate the contents of `mcoInit()` into `init.c`, and pass the parameter group array element into `void mcoConfigure(mcoConfig_t device)`.